src/backend/distributed/shared_library_init.c

```diff
@@ -924,6 +925,63 @@ RegisterCitusConfigVariables(void)
 		0,
 		NULL, NULL, NULL);
 
+	DefineCustomBoolVariable(
+		"citus.check_available_space_before_move",
+		gettext_noop("When enabled will check free disk space before a shard move"),
+		gettext_noop(
+			"Free disk space will be checked when this setting is enabled before each shard move."
+			"See citus.delete_old_shards_sleep_time and citus.delete_old_shards_max_tries"),
+		&CheckAvailableSpaceBeforeMove,
+		true,
+		PGC_USERSET,
+		0,
+		NULL, NULL, NULL);
+
+	DefineCustomIntVariable(
+		"citus.delete_old_shards_max_tries",
+		gettext_noop(
+			"Sets how many times deleting old shards will be retried before moving a shard"),
+		gettext_noop(
+			"When shard deletion is deferred, old shards will be there using some storage."
+			"When moving a shard, citus will try to delete the old shards to create "
+			"some more free space so that there is enough space for the shard move."
+			"This setting determines how many times the deletion operation will be tried."
+			"In between the tries, there will be delete_old_shards_sleep_time second sleeps."),
+		&WaitForDeferShardsMaxTries,
+		5, 0, INT_MAX,
+		PGC_USERSET,
+		0,
+		NULL, NULL, NULL);
+
+	DefineCustomIntVariable(
+		"citus.delete_old_shards_sleep_time",
+		gettext_noop(
+			"Sets sleep time between old shard delete operations prior to shard move"),
+		gettext_noop(
+			"When shard deletion is deferred, old shards will be there using some storage."
+			"When moving a shard, citus will try to delete the old shards to create "
+			"some more free space so that there is enough space for the shard move."
+			"This setting determines the duration of sleep (in second) between the deletion operations."),
+		&WaitForDeferShardRetryTimeInSec,
+		30, 0, INT_MAX,
+		PGC_USERSET,
+		GUC_UNIT_S,
+		NULL, NULL, NULL);
+
+	DefineCustomRealVariable(
+		"citus.desired_percent_disk_available_after_move",
+		gettext_noop(
+			"Sets how many percentage of free disk space should be after a shard move"),
+		gettext_noop(
+			"This setting controls how much free space should be available after a shard move."
+			"If the free disk space will be lower than this parameter, then shard move will result in"
+			"an error."),
+		&DesiredPercentFreeAfterMove,
+		5.0, 0.0, 100.0,
```

Comment:
As far as I remember, HSC converts the system to read-only if disk space left is less than 5%. So, in that regard, the smallest allowed should probably be 6%. But, lets keep the smallest allowed as 0, just for emergency cases.

But, I think the default can be 10% (or even more?).

Reply:
Ping on this, I think we should consider making the default at least 10.